### DISCUSSION TITLE
Laravel 9 Attributes how to do type check?

### DISCUSSION THREAD

**Discussion by egege** [larastan/larastan#1898] [Q&A]
Following the link below，I have written some code, but it seems that it doesn't check the return types of the get or set functions for Attributes.
> PHP Version: 8.3.4
> Laravel Version: v11.1.0
> Larastan Version: v2.9.2
> --level used: 8

https://github.com/larastan/larastan/blob/2.x/docs/features.md
```php
/**
 * @var array<int, string>
 */
public const TYPE_MAP = [
    self::TYPE_IMAGES => 'image',
    self::TYPE_VIDEOS => 'video',
];

/**
 * typeName
 * return should be Attribute<string, never>
 * @return Attribute<string[], string>
 */
protected function typeName(): Attribute
{
    return Attribute::make(
        get: function (): string {
            return self::TYPE_MAP[$this->media_type];
        }
    );
}
```

### screenshot as follows
![截屏2024-04-02 00 04 28](https://github.com/larastan/larastan/assets/23636455/82eb81a8-cc6d-49dd-94de-b70ee11e682f)
![截屏2024-04-02 00 04 40](https://github.com/larastan/larastan/assets/23636455/aa84cb0b-db95-4fc3-903d-329484f0903f)

**Comment by szepeviktor** [comment]
Thank you for opening this.
If Larastan could read the return types from `Attribute::make` the docblock would not be necessary.

**Comment by calebdw** [comment]
It does check the types, and it is necessary to be able to assign a readable type to a database column backed attribute (accessor).

For computed properties (an attribute without a database column) you can explicitly set the type to `never`:

```php
/** @return Attribute<string[], never> */
protected function typeName(): Attribute
{
    return Attribute::make(
        get: fn (): string => self::TYPE_MAP[$this->media_type],
    );
}
```

or use the `::get` method:

```php
/** @return Attribute<string[], never> */
protected function typeName(): Attribute
{
    return Attribute::get(
        fn (): string => self::TYPE_MAP[$this->media_type],
    );
}
```

There's nothing to be done here

**Comment by szepeviktor** [comment]
@egege means that the type in docblock and in `get:` are not compared.

**Reply by calebdw** [reply]
Ah my bad! I thought the `never` type was referred to. It looks like the stubs generics are set correctly but perhaps something in the code is overriding it

**Reply by egege** [reply]
Yes, that's what I mean. So it's a bug, not a feature that needs to be added?
